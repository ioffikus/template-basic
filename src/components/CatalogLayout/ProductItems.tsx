import { RouterState } from 'connected-react-router';
import { connect } from 'react-redux';
import ProductItems from 'src/components/ProductItems';

interface IOwnProps {}

interface IConnectedState {
  router: RouterState;
  ui: Alicanto.Models.UI;
  isRequesting: boolean;
  items: Alicanto.Models.CatalogItem[];
  hasNextPage: boolean;
}

const mapStateToProps = (state: Store.IState) => ({
  router: state.router,
  ui: state.ui,
  isRequesting: state.catalog.isRequesting,
  items: state.catalog.results,
  hasNextPage: !!state.catalog.next,
});

const WrappedProductItems = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(ProductItems);

export default WrappedProductItems;
