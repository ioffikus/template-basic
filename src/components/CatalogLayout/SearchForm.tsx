import React from 'react';
import { RouterState, push } from 'connected-react-router';
import { connect } from 'react-redux';
import { Form, Input, Icon, Button } from 'antd';
import cx from 'classnames';
import queryString from 'query-string';
import dispatch from 'src/store/dispatch';
import Svg from 'src/components/Svg';

interface IOwnProps {}

interface IConnectedState {
  router: RouterState;
  search: string;
  ui: Alicanto.Models.UI;
}

type IProps = IOwnProps & IConnectedState;

interface IState {
  value: string;
  isOpen: boolean;
}

const mapStateToProps = (state: Store.IState) => ({
  router: state.router,
  search: state.catalogQuery.search,
  ui: state.ui,
});

const WrappedSearchForm = connect<IConnectedState, undefined, IOwnProps>(mapStateToProps)(
  class SearchForm extends React.Component<IProps, IState> {
    state: IState = {
      value: this.props.search,
      isOpen: false,
    };

    goToSearch = (search: string) => {
      const query = {
        ...queryString.parse(this.props.router.location.search),
        search,
      };
      delete query.page;
      const location = { ...this.props.router.location, search: queryString.stringify(query) };
      dispatch(push(location));
    };

    handleToggle = () => {
      if (this.state.isOpen) {
        this.setState({
          value: '',
          isOpen: false,
        });
        if (this.props.search) {
          this.goToSearch('');
        }
      } else {
        this.setState({
          isOpen: true,
        });
      }
    };

    handleChange = (event: React.SyntheticEvent<{ value: string }>) => {
      const value = event.currentTarget.value;
      this.setState({
        value,
      });
    };

    handlePressEnter = () => {
      const search = this.state.value;
      this.goToSearch(search);
    };

    render() {
      const catalogSearchFormUI = this.props.ui.common.catalogSearchForm;
      return (
        <div
          className={cx('root', {
            open: this.state.isOpen,
          })}
        >
          <style jsx>{`
            .root {
              display: flex;
              align-items: center;
              position: relative;
              margin-left: auto;
              width: 32px;
              height: 32px;
            }
            .inner {
              position: absolute;
              right: 0;
              top: 50%;
              z-index: 2;
              transform: translateY(-50%);
              transition: width 0.15s cubic-bezier(0.645, 0.045, 0.355, 1), opacity 2.5s;
              opacity: 0;
              width: 32px;
              margin-right: 2px;
              pointer-events: none;
            }
            .open :global(.inner) {
              opacity: 1;
              width: 440px;
              pointer-events: auto;
              transition: width 0.15s cubic-bezier(0.645, 0.045, 0.355, 1), opacity 0s;
            }
            .root :global(.ant-form-item),
            .root :global(.ant-form-item-control-wrapper) {
              width: 100%;
              margin-right: 0;
            }
            .root :global(.ant-input) {
              height: 32px;
              padding-left: 0;
              padding-right: 0;
              border-radius: 100px;
              background-color: #e5ecf4;
              color: #b2bdcb;
              border: 0;
              box-shadow: none;
              font-size: 12px;
              transition: none;
            }
            .open :global(.ant-input) {
              padding-left: 16px;
              padding-right: 0;
            }
            .root :global(.ant-input::placeholder) {
              opacity: 0;
              transition: none;
            }
            .open :global(.ant-input::placeholder) {
              opacity: 1;
              transition: 0.15s;
            }
            .root :global(.ant-input-suffix) {
              color: #113851;
            }
            .root :global(.ant-input-suffix) {
              right: 11px;
              opacity: 0;
              backface-visibility: hidden;
              -webkit-font-smoothing: antialiased;
            }
            .open :global(.ant-input-suffix) {
              opacity: 1;
              transition: 0.3s;
            }
            .root :global(.anticon-search) {
              font-size: 16px;
            }
            .root :global(.anticon-close) {
              font-size: 14px;
              font-weight: bold;
              cursor: pointer;
              backface-visibility: hidden;
              -webkit-font-smoothing: antialiased;
            }
            .root :global(.icon-search) {
              background: transparent;
            }
            .root :global(.icon-search:after) {
              display: none;
            }
            .root :global(.icon-svg) {
              width: 16px;
              height: 16px;
            }
          `}</style>
          <div className="inner">
            <Form layout="inline">
              <Form.Item>
                <Input
                  placeholder={catalogSearchFormUI.placeholder.i18n.defaultText}
                  suffix={<Icon type="close" onClick={this.handleToggle} />}
                  value={this.state.value}
                  onChange={this.handleChange}
                  onPressEnter={this.handlePressEnter}
                />
              </Form.Item>
            </Form>
          </div>
          <Button className="icon-search" shape="circle" onClick={this.handleToggle}>
            <Svg name="search" className="icon-svg" />
          </Button>
        </div>
      );
    }
  },
);

export default WrappedSearchForm;
