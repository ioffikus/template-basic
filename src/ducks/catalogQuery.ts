import { createAction, createReducer } from 'redux-act';
import queryString from 'query-string';
import _ from 'lodash';
import { safelyParseJSON } from 'src/helpers/safelyParseJSON';
import { IObject } from 'src/core/interfaces/IObject';

export const REDUCER = 'CATALOG_QUERY';
const NS = `${REDUCER}__`;

export interface ICatalogQueryState {
  search: string;
  page: string;
  pageSize: string;
  ordering: string;
  filterObj: IObject<Alicanto.API.CatalogFilter>;
}

export const initialState: ICatalogQueryState = {
  search: '',
  page: '1',
  pageSize: '12',
  ordering: '',
  filterObj: {},
};

const reducer = createReducer<ICatalogQueryState>({}, initialState);

const set = createAction<ICatalogQueryState>(`${NS}SET`);
reducer.on(set, (state, { search, page, pageSize, ordering, filterObj }) => ({
  ...state,
  search,
  page,
  pageSize,
  ordering,
  filterObj: { ...filterObj },
}));

export const getCatalogQuery = (): Store.IReduxThunkCb => (dispatch, getState) => {
  const router = getState().router;
  const query = queryString.parse(router.location.search);
  const search = _.get(query, 'search', initialState.search);
  const page = _.get(query, 'page', initialState.page);
  const pageSize = _.get(query, 'page_size', initialState.pageSize);
  const ordering = _.get(query, 'ordering', initialState.ordering);
  const filterStr = _.get(query, 'filter', null);
  const filterObj = safelyParseJSON(filterStr) || {};
  dispatch(set({ search, page, pageSize, ordering, filterObj }));
  return Promise.resolve();
};

export default reducer;
