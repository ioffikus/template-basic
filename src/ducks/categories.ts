import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CATEGORIES';
const NS = `${REDUCER}__`;

export interface ICategoriesState extends Alicanto.API.GET_CATALOG_CATEGORIES_RESPONSE, Store.IReduxCommonState {}

export const initialState: ICategoriesState = {
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<ICategoriesState>({}, initialState);

const callSuccess = createAction<Alicanto.API.GET_CATALOG_CATEGORIES_RESPONSE>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { next, previous, count, results }) => ({
  ...state,
  next,
  previous,
  count,
  results: [...results],
  isRequestError: false,
  isRequesting: false,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequesting: false,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

export const getCategories = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: EndpointsConfig.CATEGORIES,
        params: { page_size: '100' },
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CATALOG_CATEGORIES_RESPONSE;
        dispatch(callSuccess(data));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
