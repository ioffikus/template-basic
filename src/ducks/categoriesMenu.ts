import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CATEGORIES_MENU';
const NS = `${REDUCER}__`;

interface ICategoriesMenuResponse extends Alicanto.API.GET_CATALOG_GROUPS_SLUG_RESPONSE {
  content: Alicanto.Models.Category[];
}

export interface ICategoriesMenuState extends ICategoriesMenuResponse, Store.IReduxCommonState {}

export const initialState: ICategoriesMenuState = {
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<ICategoriesMenuState>({}, initialState);

const callSuccess = createAction<ICategoriesMenuResponse>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { slug, type, content }) => {
  return {
    ...state,
    slug,
    type,
    content: [...content],
    isRequestError: false,
    isRequesting: false,
  };
});

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  slug: '',
  type: '',
  content: [],
  isRequesting: false,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

export const getCategoriesMenu = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    GET(
      {
        url: EndpointsConfig.CATEGORIES_MENU,
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as ICategoriesMenuResponse;
        dispatch(callSuccess(data));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
