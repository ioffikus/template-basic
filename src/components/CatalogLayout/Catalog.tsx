import React from 'react';
import { Row, Col } from 'antd';
import Breadcrumbs from 'src/components/Breadcrumbs';
import SearchFrom from './SearchForm';
import ProductItems from './ProductItems';

class Catalog extends React.Component<{}, {}> {
  render() {
    return (
      <div className="root app-container">
        <style jsx>{`
          .root :global(.header) {
            margin-bottom: 30px;
          }
        `}</style>
        <Row type="flex" justify="space-between" className="header">
          <Col span={23}>
            <Breadcrumbs for="catalog" />
          </Col>
          <Col span={1}>
            <SearchFrom />
          </Col>
        </Row>
        <ProductItems />
      </div>
    );
  }
}

export default Catalog;
