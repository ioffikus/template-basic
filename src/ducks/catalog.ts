import { createAction, createReducer } from 'redux-act';
import { AxiosResponse } from 'axios';
import _ from 'lodash';
import { ICatalogQueryState } from 'src/ducks/catalogQuery';
import { GET } from 'src/core/API';
import EndpointsConfig from 'src/configs/endpoints';

export const REDUCER = 'CATALOG';
const NS = `${REDUCER}__`;

export interface ICatalogState extends Alicanto.API.GET_CATALOG_RESPONSE, Store.IReduxCommonState {}

export const initialState: ICatalogState = {
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequesting: false,
  isRequestError: false,
};

const reducer = createReducer<ICatalogState>({}, initialState);

const callSuccess = createAction<Alicanto.API.GET_CATALOG_RESPONSE>(`${NS}CALL_SUCCESS`);
reducer.on(callSuccess, (state, { next, previous, count, results }) => ({
  ...state,
  next,
  previous,
  count,
  results: [...results],
  isRequesting: false,
  isRequestError: false,
}));

const callFailure = createAction(`${NS}CALL_FAILURE`);
reducer.on(callFailure, state => ({
  ...state,
  next: null,
  previous: null,
  count: 0,
  results: [],
  isRequesting: false,
  isRequestError: true,
}));

const callRequest = createAction(`${NS}CALL_REQUEST`);
reducer.on(callRequest, state => ({
  ...state,
  isRequesting: true,
}));

interface ICatalogQuery {
  search?: string;
  page?: string;
  page_size: string;
  ordering?: string;
  filter?: string;
}

const getCatalogRequestParams = ({ search, page, pageSize, ordering, filterObj }: ICatalogQueryState) => {
  const params: ICatalogQuery = {
    page_size: pageSize,
  };
  if (!_.isEmpty(search)) {
    params.search = search;
  }
  if (!_.isEmpty(page)) {
    params.page = page;
  }
  if (!_.isEmpty(ordering)) {
    params.ordering = ordering;
  }
  if (!_.isEmpty(filterObj)) {
    params.filter = JSON.stringify(filterObj);
  }
  return params;
};

export const getCatalog = (abortToken: string): Store.IReduxThunkCb => (dispatch, getState, xShopHeaders) => {
  return new Promise((resolve, reject) => {
    dispatch(callRequest());
    const catalogQuery = getState().catalogQuery;
    GET(
      {
        url: EndpointsConfig.CATALOG,
        params: getCatalogRequestParams(catalogQuery),
      },
      xShopHeaders,
      abortToken,
    )
      .then((resp: AxiosResponse) => {
        const data = resp.data as Alicanto.API.GET_CATALOG_RESPONSE;
        let results = data.results;
        if (catalogQuery.page !== '1') {
          const oldResults = getState().catalog.results;
          results = [...oldResults, ...results];
        }
        dispatch(callSuccess({ ...data, results }));
        resolve();
      })
      .catch(error => {
        dispatch(callFailure());
        reject(error);
      });
  });
};

export default reducer;
