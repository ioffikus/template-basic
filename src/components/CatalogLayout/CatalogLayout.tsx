import React from 'react';
import { Layout } from 'antd';
import dispatch from 'src/store/dispatch';
import { getCatalogQuery } from 'src/ducks/catalogQuery';
import { getCategoriesMenu } from 'src/ducks/categoriesMenu';
import { getCatalog } from 'src/ducks/catalog';
import { getCategories } from 'src/ducks/categories';
import { getFilter } from 'src/ducks/filter';
import SiderMenu from 'src/components/SiderMenu';
import CategoriesMenu from 'src/components/CategoriesMenu';
import FilterMenu from 'src/components/FilterMenu';
import Catalog from './Catalog';

class CatalogLayout extends React.Component<{}, {}> {
  static fetchData: Store.IFetchDataFn = abortToken => {
    return dispatch(getCatalogQuery()).then(() => {
      return Promise.all([
        dispatch(getCategoriesMenu(abortToken)),
        dispatch(getCatalog(abortToken)),
        dispatch(getCategories(abortToken)),
        dispatch(getFilter(abortToken)),
      ]);
    });
  };

  render() {
    return (
      <Layout>
        <SiderMenu type="categories">
          <CategoriesMenu />
        </SiderMenu>
        <Layout.Content>
          <Catalog />
        </Layout.Content>
        <SiderMenu type="filters">
          <FilterMenu />
        </SiderMenu>
      </Layout>
    );
  }
}

export default CatalogLayout;
