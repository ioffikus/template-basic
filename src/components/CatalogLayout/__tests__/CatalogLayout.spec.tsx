import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import nanoid from 'nanoid';
import { setupDispatch } from 'src/store/dispatch';
import { getCatalogQuery } from 'src/ducks/catalogQuery';
import { getCategoriesMenu } from 'src/ducks/categoriesMenu';
import { getCatalog } from 'src/ducks/catalog';
import { getCategories } from 'src/ducks/categories';
import { getFilter } from 'src/ducks/filter';
import CatalogLayout from '../CatalogLayout';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
jest.mock('src/ducks/catalogQuery', () => {
  return {
    getCatalogQuery: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});
jest.mock('src/ducks/categoriesMenu', () => {
  return {
    getCategoriesMenu: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});
jest.mock('src/ducks/catalog', () => {
  return {
    getCatalog: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});
jest.mock('src/ducks/categories', () => {
  return {
    getCategories: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});
jest.mock('src/ducks/filter', () => {
  return {
    getFilter: jest.fn(() => {
      return () => Promise.resolve();
    }),
  };
});

describe(null, () => {
  it('fetchData()', () => {
    const store = mockStore({});
    setupDispatch({ store });
    const abortToken = nanoid();
    return CatalogLayout.fetchData(abortToken).then(data => {
      expect(getCatalogQuery).toBeCalled();
      expect(getCategoriesMenu).toBeCalledWith(abortToken);
      expect(getCatalog).toBeCalledWith(abortToken);
      expect(getCategories).toBeCalledWith(abortToken);
      expect(getFilter).toBeCalledWith(abortToken);
    });
  });
  it('render()', () => {
    const store = mockStore({});
    const wrapper = shallow(<CatalogLayout />, { context: { store } });
    expect(wrapper).toMatchSnapshot();
  });
});
