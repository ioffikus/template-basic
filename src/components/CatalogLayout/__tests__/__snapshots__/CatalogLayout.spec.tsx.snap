// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render() 1`] = `
<Adapter>
  <SiderMenu
    type="categories"
  >
    <Connect(CategoriesMenu) />
  </SiderMenu>
  <Adapter>
    <Catalog />
  </Adapter>
  <SiderMenu
    type="filters"
  >
    <Connect(FilterMenu) />
  </SiderMenu>
</Adapter>
`;
